hibernate-reactive-core/src/main/java/org/hibernate/reactive/engine/impl/CollectionTypes.java

```diff
@@ -0,0 +1,481 @@
+		return loop(
+				0, length, i -> {
+					return getReplace( elemType, Array.get( original, i ), owner, session, copyCache )
+							.thenCompose( o -> {
+											  Array.set( result, i, o );
+											  return completedFuture( result );
+										  }
+							);
+				}
+		).thenApply( unused -> result );
+	}
+
+	private static CompletionStage<Object> getReplace(
+			Type elemType,
+			Object o,
+			Object owner,
+			SessionImplementor session,
+			Map<Object, Object> copyCache) {
+		return getReplace( elemType, o, null, owner, session, copyCache );
+	}
+
+	private static CompletionStage<Object> getReplace(
+			Type elemType,
+			Object o,
+			Object target,
+			Object owner,
+			SessionImplementor session,
+			Map<Object, Object> copyCache) {
+		if ( elemType instanceof EntityType ) {
+			return EntityTypes.replace( (EntityType) elemType, o, target, session, owner, copyCache );
+		}
+		else {
+			final Object replace = elemType.replace( o, target, session, owner, copyCache );
+			return completedFuture( replace );
+		}
+	}
+
+	private static CompletionStage<Void> preserveSnapshot(
+			PersistentCollection<?> original,
+			PersistentCollection<?> result,
+			Type elemType,
+			Object owner,
+			Map<Object, Object> copyCache,
+			SessionImplementor session) {
+		final CollectionEntry ce = session.getPersistenceContextInternal().getCollectionEntry( result );
+		if ( ce != null ) {
+			return createSnapshot( original, result, elemType, owner, copyCache, session )
+					.thenCompose( CompletionStages::completedFuture )
+					.thenAccept( serializable ->
+										 ce.resetStoredSnapshot( result, serializable ) );
+		}
+		return voidFuture();
+	}
+
+	private static CompletionStage<Serializable> createSnapshot(
+			PersistentCollection<?> original,
+			PersistentCollection<?> result,
+			Type elemType,
+			Object owner,
+			Map<Object, Object> copyCache,
+			SessionImplementor session) {
+		final Serializable originalSnapshot = original.getStoredSnapshot();
+		if ( originalSnapshot instanceof List<?> list ) {
+			return createListSnapshot( list, elemType, owner, copyCache, session );
+		}
+		else if ( originalSnapshot instanceof Map<?, ?> map ) {
+			return createMapSnapshot( map, result, elemType, owner, copyCache, session );
+		}
+		else if ( originalSnapshot instanceof Object[] array ) {
+			return createArraySnapshot( array, elemType, owner, copyCache, session );
+		}
+		else {
+			// retain the same snapshot
+			return completedFuture( result.getStoredSnapshot() );
+		}
+	}
+
+	private static CompletionStage<Serializable> createArraySnapshot(
+			Object[] array,
+			Type elemType,
+			Object owner,
+			Map<Object, Object> copyCache,
+			SessionImplementor session) {
+		return loop(
+				0, array.length,
+				i ->
+						getReplace( elemType, array[i], owner, session, copyCache )
+								.thenCompose( o -> {
+									array[i] = o;
+									return voidFuture();
+								} ).thenApply( unused -> voidFuture() )
```

Comment:
You shouldn't return `.voidFuture()` when using `.thenApply`
In any case, you can replace this
```java
.thenCompose( o -> {
    array[i] = o;
    return voidFuture();
} ).thenApply( unused -> voidFuture() )
```
with
```java
.thenCompose( o -> {
    array[i] = o;
} )
```